[12] Fetch Google patents

beingkk opened the pull request:
Closes #4 and #12 

Adds a refactored pipeline to fetch Google Patents based on a set of keywords.
- The fetching pipeline also expands the keywords by substituting a few synonyms (child = baby = infant = toddlers)
- I have also included options for doing a "dry run" and checking the how much data will be scanned by the query (directly proportional to the cost of the query)
- The queried data is timestamped (to avoid overwriting results) and saved on s3. In addition, I'm also saving for reference the keywords and the exact query used for fetching the data from BigQuery.

There is also a notebook (`00_fetch_data`) that could be helpful in iterating and improving the query in the future.

I also refactored a module to deal with keywords separately from the google_utils, as I think they'll be useful for other datasets as well.

For transparency, the keywords themselves are stored in the config folder and visible on GitHub 

# Instructions for Reviewer

If you'd like, you can also run the query - note that costs us around ~£1 (one pound) per run, but would be good to try at least once to check that you can reproduce.
```python
python discovery_child_development/pipeline/patents/fetch_patents.py
```

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

beingkk committed:
delete the expensive query

beingkk committed:
testing logic mill

beingkk committed:
simple example of using logic mill

RFOxbury committed:
WIP: preparing training data

RFOxbury committed:
Utilise nesta_ds_utils and remove boto3 dependency

RFOxbury committed:
Add labelling utils from innovation_sweet_spots

RFOxbury committed:
WIP: prepare training data and start using wandb

RFOxbury committed:
Refactor a bit

RFOxbury committed:
Train a classifer and log this on wandb

RFOxbury committed:
Add figures to gitignore

RFOxbury committed:
Add process for training dummy baseline classifier

beingkk committed:
fetching patents

RFOxbury committed:
WIP: Create module for running baseline model

RFOxbury committed:
Refactor slightly and add a test

beingkk committed:
adjusting requirements and env_template

RFOxbury committed:
Merge branch 'dev' into 15-train-baseline-model

beingkk committed:
fetching child development patents

beingkk committed:
delete stray figure

RFOxbury committed:
Merge branch 'dev' into 15-train-baseline-model

RFOxbury committed:
Fix issue with probability calculation in random choice classifier

RFOxbury committed:
Add comments and small correction

beingkk committed:
Merge branch 'dev' into 12_google_patents

RFOxbury committed:
Add docstrings and some tests

RFOxbury committed:
Remove unusued libraries

beingkk committed:
clustering and summarising

beingkk committed:
refactor querying

beingkk committed:
refactor querying

beingkk committed:
Merge branch '15-train-baseline-model' into 12_google_patents

beingkk committed:
renaming bigquery to google_utils

beingkk commented:
FYI @tomwillcocks1 if you're curious - here's some code that uses SQL to access a large patent dataset on Google's BigQuery. No need to review, but just for your knowledge - and can walk you through on Monday if that's interesting.

beingkk committed:
fix dependencies

RFOxbury commented on discovery_child_development/utils/google_utils.py:
```diff
@@ -79,3 +84,125 @@ def create_client() -> bigquery.Client:
+        FROM `patents-public-data.patents.publications` pub
+            INNER JOIN `patents-public-data.google_patents_research.publications` gpr ON
+            pub.publication_number = gpr.publication_number
+        WHERE
```
I think you need an extra set of parentheses here? I think at the moment there's a risk that the date range restriction only applies to the abstract condition, not the title condition. Putting both the title and abstract conditions (both sides of the OR) in parentheses should fix this.

```suggestion
        WHERE
            (({combined_conditions_title})
            OR ({combined_conditions_abstract}))
            AND pub.grant_date BETWEEN 20190101 AND 20231231
```

RFOxbury replied:
See also [the order in which operators are evaluated](https://cloud.google.com/bigquery/docs/reference/standard-sql/operators)

beingkk replied:
Good catch, thank you!

RFOxbury commented on discovery_child_development/pipeline/patents/fetch_patents.py:
```diff
@@ -0,0 +1,42 @@
+    query = google_utils.create_patents_query(keywords)
+    # Run the query
+    client = google_utils.create_client()
+    google_utils.dry_run(client, query)
```
Is it worth either adding a `time.sleep()` or requiring a user input here? I am thinking of how to use `dry_run()` so that if from the logs the query looked much bigger than you were expecting, you would have the option to kill the process 😅

beingkk replied:
Nice idea - I'm thinking as this is a pipeline script, the user has already checked that the query is OK and implemented it in the pipeline. The dry run here is hence just for reference. But you're right, we could add a user prompt - will add an enhancement issue!

RFOxbury commented on discovery_child_development/utils/google_utils.py:
```diff
@@ -79,3 +84,125 @@ def create_client() -> bigquery.Client:
+        SELECT DISTINCT
+            pub.publication_number
+        FROM `patents-public-data.patents.publications` pub
+            INNER JOIN `patents-public-data.google_patents_research.publications` gpr ON
```
Would you be able to document somewhere why it's necessary to inner join these specific tables - why we need only publications that have appeared in both tables?

Is it the case that the `google_patents_research.publications` table has abstracts translated into English, plus embeddings if we want to use them, while `patents.publication` has the grant date?

beingkk replied:
Yes that's the reason - I'll add it in the docstring

RFOxbury reviewed (changes requested):
Hey Karlis, thanks so much for this PR and for figuring out the patents data! Seems like accessing patents is much more complicated than getting the OpenAlex data 🙈

I have:

- read through the keywords `.txt` files and the utils functions in `utils/keywords.py`
- read through the additional functions in `google_utils.py`
- Run the notebook `00_fetch_data.py` interactively

I have not yet run the script `pipeline/patents/fetch_patents.py` as I think that will pull more data compard to the notebook (because of the additional "toddler" term) and I did not want to incur extra costs.

I think a small fix to the SQL query is needed because currently it brings back patents from outside the specified date range (see screenshot). I think this can be solved by putting the `OR` part of the `WHERE` clause in parentheses so that the `AND` part applies to both sides of the `OR`.

![Screenshot 2023-11-03 at 18 04 41](https://github.com/nestauk/discovery_child_development/assets/49554920/adb21948-74d8-4c35-9f51-73e02a549d89)

RFOxbury commented on discovery_child_development/utils/google_utils.py:
```diff
@@ -79,3 +84,125 @@ def create_client() -> bigquery.Client:
+    return q
+
+
+def dry_run(client: bigquery.Client, query: str) -> None:
```
This function is a really good idea!

RFOxbury commented on discovery_child_development/utils/google_utils.py:
```diff
@@ -79,3 +84,125 @@ def create_client() -> bigquery.Client:
+    query: str,
+    metadata: List[str] = None,
+) -> None:
+    """Upload query results to S3. Saves the results in a subfolder
```
Really good idea to timestamp the data... Should I make an issue to do the same for OpenAlex?

beingkk replied:
Making an issue sounds like a good idea! Overall, I think lower priority than getting on with the classification algos, but can come back to it at some point.

RFOxbury commented on discovery_child_development/utils/keywords.py:
```diff
@@ -0,0 +1,49 @@
+
+
+def get_keywords(file: str) -> List[List[str]]:
+    """Get keywords from a txt file. The keywords stored in a text file as
```
Thanks for the clear docstrings! 🙏🏻

beingkk committed:
Merge branch 'dev' into 12_google_patents

beingkk commented:
Thank you very much @RFOxbury for the thorough review - and spotting the issue with the query! I've not adjusted as you suggest, and I've also added a filter to include patents that have been not just granted in the last five years, but also filed.

And then I've adjusted the query further to also include other metadata like country codes and inventors.

Thanks!

beingkk committed:
fixing query
